### PR TITLE
adds fuzzy_search algo

### PR DESCRIPTION
**Pull request by gang-qiu**

### CODE REVIEW

File: src/code-tabs.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { TabContent, Nav } from 'reactstrap';
+import CodeTabNav from './code-tabs-nav';
+import CodeTabPane from './code-tab-pane';
+
+export default class CodeTabs extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.toggle = this.toggle.bind(this);
+    this.state = {
```
Comment (daniellacosse) [inline src/code-tabs.js]: Like we discussed, needs to be moved into the App container.

Btw, functional components are a thing. You can just do
```
({ children }) => <div>{children}</div>
```
and react will know what you mean. These components are stateless, which helps enforce this idea of migrating state up

File: src/code-tabs.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      let tabIndex = parseInt(tabId, 10) - 1; 
+      let tabs = this.state.tabs;
+      tabs[tabIndex].code = code;
+      this.setState({tabs: tabs});
+      console.log(this.state.tabs);
+    } 
+  }
+
+  render() {
+    return (
```
Comment (daniellacosse) [inline src/code-tabs.js]: feel free to `const {tabs} = this.state` here

File: src/fuzzy_search.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+
+
+// fuzzy search lookup via decomposition of characters in a word
+// provides constant time lookup
+
+class Node {
```
Comment (daniellacosse) [inline src/fuzzy_search.js]: Let's move this to a different pr

File: src/code-tabs-nav.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+
+export default class CodeTabNav extends React.Component {
+  // nest this under <Nav tabs> bootstrap element
+  render() {
+    return (
+      <NavItem>
+        <NavLink
+          className={classnames({ active: this.props.active })}
+          onClick={() => { this.props.onClick()}}
```
Comment (daniellacosse) [inline src/code-tabs-nav.js]: don't need to wrap this in a function, just pass `this.props.onClick`

File: src/code-tabs.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  toggle(tab) {
+    if (this.state.activeTab !== tab) {
+      this.setState({
+        activeTab: tab
+      });
+    }
+  }
+
+  addNewTab() {
+    let lastTab = this.state.tabs[this.state.tabs.length - 1]
```
Comment (daniellacosse) [inline src/code-tabs.js]: would be a good place to de-structure `tabs` out

File: src/code-tab-pane.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import { TabPane } from 'reactstrap';
+import CodeMirror from 'react-codemirror';
+
+export default class CodeTabPane extends React.Component {
+  // nest this under <Nav tabs> bootstrap element
+  render() {
+    return (
+      <TabPane tabId={this.props.id}>
+        <CodeMirror onChange={this.props.onChange}/>
```
Comment (daniellacosse) [inline src/code-tab-pane.js]: There're a couple more props you need to pass into `CodeMirror`, like `options`, to get the syntax highlighting and stuff. [Refer to this](https://github.com/daniellacosse/js-profiler/blob/master/src/App.js#L93)

File: public/index.html [hunk 7]
```diff
@@ -21,7 +21,7 @@
     -->
     <title>React App</title>
   </head>
-  <body>
+  <body class="container">
```
Comment (daniellacosse) [inline public/index.html]: I'd put this on the `<main>` tag in `App.js`

File: src/code-tabs.js (hunk 5 above)
Comment (gang-qiu) [inline src/code-tabs.js]: not sure how to destructure this

**Comment by gang-qiu** [comment]
moved up state and did some refactoring

File: src/code-tabs.js (hunk 5 above)
Comment (daniellacosse) [inline src/code-tabs.js]: a-like
```
const {tabs} = this.state;
let lastTab = tabs[tabs.length -1]a
```
and such forth
